### ISSUE TITLE
"Show whitespace" should also show "Zero Width Whitepace"

### ISSUE THREAD

**Issue by BeckerWdf** [eclipse-platform/eclipse.platform.ui#1002]
Source Code (that e.g. was copied from a web page) might contain zero width whitespaces (ZWSP). That this is can be read here;
https://en.wikipedia.org/wiki/Zero-width_space#:~:text=The%20zero%2Dwidth%20space%20(%E2%80%8B,but%20after%20which%20there%20may

The "Show Whiteplace" function of the text editor does show whitespaces and line breaks etc. but does not show ZWSPs.

Other editor (e.g. Notepad++ on windows) can show this. See:
![telegram-cloud-photo-size-2-5379835439394965993-m](https://github.com/eclipse-platform/eclipse.platform.ui/assets/28338612/808e04db-fab4-4049-8660-6be61b1d92e1)

I think eclipse should also be capable of showing this.

If you are asking how to enter an ZWSP into your editor: Use this website: https://zerowidthspace.me/

**Comment by BeckerWdf** [comment]
This was already proposed in https://bugs.eclipse.org/bugs/show_bug.cgi?id=578213

**Comment by BeckerWdf** [comment]
org.eclipse.jface.text.WhitespaceCharacterPainter is the place where the whitespace char drawing happens

**Comment by lathapatil** [comment]
> This was already proposed in https://bugs.eclipse.org/bugs/show_bug.cgi?id=578213

Can this enhancement to show invisible characters be extended to all Zero width characters as mentioned here ?https://plugins.jetbrains.com/plugin/7448-zero-width-characters-locator

**Comment by marcushoepfner** [comment]
Would you like to see this as an option here?

<img width="706" alt="show_whitespace_characters" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/29301444/ac28895b-7f7d-41b7-b7ed-a43beb685337">

**Comment by lathapatil** [comment]
> Would you like to see this as an option here?


Yes. Ideally ZWSP proposed  by author might come here . Along with this other Zero width characters can be implemented as part of this enhancement.

**Comment by marcushoepfner** [comment]
The responsible java class is `WhitespaceCharacterPainter` (method `drawCharRange`).
It seems, that it replaces one character by another, e.g. space by middle dot.

I tried to replace a zero with white space with the letters "ZWSP". But then it looks strange when I continue typing.

This is only the ZWSP pasted from https://zerowidthspace.me/ (it has a width of 0 of course). It looks ok:
<img width="218" alt="zwsp" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/29301444/0b6ce0b0-7483-4815-8d98-11efdaa820e4">
But when typing is continued, it looks strange:
<img width="212" alt="zwsp_typing_continued" src="https://github.com/eclipse-platform/eclipse.platform.ui/assets/29301444/1e91f58f-b47b-43c5-9790-a9888f3e9710">

**Comment by lathapatil** [comment]
ZWSP is invisible space and I have pasted it between letter "e" and "."  and Show whitespace characters is **off** in preferences.

![zwsp_prefereneceOff](https://github.com/eclipse-platform/eclipse.platform.ui/assets/130035155/dad378ad-805f-4f56-99f6-20b88171493f)

I tried to modify the code so that ZWSP is replaced with \u2588 , but it is drawn on next character as below.. when Show whitespace characters is **on** in preferences.

![ZWSP_Preference_ON](https://github.com/eclipse-platform/eclipse.platform.ui/assets/130035155/a6580975-85f4-43c9-8603-d1472d29c650)

Currently Eclipse is supporting all visible characters so that there is space and where new symbol can be drawn but for invisible characters there is no space to draw new symbol .
For CR and LF , when text is read to load itself it is read as \r and \n so that it can be easily drawn in this place.

Should we push next available character to make space for it ? so that drawing happens on later stage.. But there are challenges on how to achieve this !

**Comment by BeckerWdf** [comment]
> Should we push next available character to make space for it ? so that drawing happens on later stage.. But there are challenges on how to achieve this !

How do other editors handle that challenges?

**Comment by BeckerWdf** [comment]
> Should we push next available character to make space for it ? so that drawing happens on later stage.

Maybe we could do it similar like it's done for the code mining. For code minings also content is drawn for which we need to make space in the editor. The difference is that for the code minings we need additional vertical space for the none-width chars we need additional horizontal space.

**Comment by BeckerWdf** [comment]
> The difference is that for the code minings we need additional vertical space for the none-width chars we need additional horizontal space.

That's not true. See for example [InlinedAnnotationDemo.java](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/examples/org.eclipse.jface.text.examples/src/org/eclipse/jface/text/examples/sources/inlined/InlinedAnnotationDemo.java) from the jface examples. Here the strings "red:", "green:" and "blue:" are drawn inline into the text.

**Comment by BeckerWdf** [comment]
> > The difference is that for the code minings we need additional vertical space for the none-width chars we need additional horizontal space.
> 
> That's not true. See for example [InlinedAnnotationDemo.java](https://github.com/eclipse-platform/eclipse.platform.ui/blob/master/examples/org.eclipse.jface.text.examples/src/org/eclipse/jface/text/examples/sources/inlined/InlinedAnnotationDemo.java) from the jface examples. Here the strings "red:", "green:" and "blue:" are drawn inline into the text.

Support for this seems to be completely in jface not in SWT.

**Comment by mickaelistria** [comment]
I think you should indeed consider using InlinedAnnotation for the case of adding "artificial" content in a text line.

**Comment by marcushoepfner** [comment]
I have create a first PR.

Some questions.
1. Is the used plugin the correct sport for this?

2. There is a white space character preference page, the question is, whether we should add ZWSP?
In contradiction the white space pref page is about the show white space character functionality which is handled in org.eclipse.jface.text.WhitespaceCharacterPainter, but in my commit code mining is used. But currently there is not Code Mining preference page for text editors itself, just for java editors.

**Comment by mickaelistria** [comment]
I don't think a new pref is necessary. Relying on the exisying Show whitespaces is enough

**Comment by BeckerWdf** [comment]
> 2. There is a white space character preference page, the question is, whether we should add ZWSP?

I Would say yes.

>    In contradiction the white space pref page is about the show white space character functionality which is handled in org.eclipse.jface.text.WhitespaceCharacterPainter, but in my commit code mining is used. But currently there is not Code Mining preference page for text editors itself, just for java editors.

But isn't this just an implementation detail. From a user perspective I would say these settings belong together.

**Comment by marcushoepfner** [comment]
> I don't think a new pref is necessary. Relying on the exisying Show whitespaces is enough

Ah, the whitespace pref page is very detailed. Refer to https://github.com/eclipse-platform/eclipse.platform.ui/issues/1002#issuecomment-1787297264. So we can add a setting for ZWSP.

**Comment by BeckerWdf** [comment]
> Ah, the whitespace pref page is very detailed. Refer to [#1002 (comment)](https://github.com/eclipse-platform/eclipse.platform.ui/issues/1002#issuecomment-1787297264). So we can add a setting for ZWSP.

This preference page is located in org.eclipse.ui.editors. See;
https://github.com/eclipse-platform/eclipse.platform.ui/blob/44efc1d09a56d4c83a5f529b64aee54a757394f9/bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/editors/text/TextEditorDefaultsPreferencePage.java#L408

org.eclipse.ui.editors today has a dependency to org.eclipse.ui.workbench.texteditor. I think with this PR we would add a circular dependency when org.eclipse.ui.workbench.texteditor accesses the preference store of org.eclipse.ui.editors.

Wouldn't it be easier to move the org.eclipse.ui.workbench.texteditor.codeMiningProviders extension to the plugin.xml of org.eclipse.ui.editors?
